File: Data Structures & Algorithms/subsets/submission-2.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number[][]}
     */
    subsets(nums) {
        const ret = []
        const dfs = ((i, subset) =>{
            if(i == nums.length){
                ret.push([...subset])
                return
            }
            subset.push(nums[i])
            dfs(i + 1, subset)
            subset.pop()
            dfs(i + 1, subset)
        })
        dfs(0, [])
        return ret
    }
}

File: Data Structures & Algorithms/generate-parentheses/submission-2.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        ret = []
        def backtrack(openB, closedB, subset):
            nonlocal ret
            if(openB == closedB == n):
                # print("".join(subset))
                ret.append("".join(subset))
                return
            
            if(openB < n):
                subset.append("(")
                print(subset)
                backtrack(openB + 1, closedB, subset)
                subset.pop()
            
            if(openB > closedB):
                subset.append(")")
                backtrack(openB, closedB + 1, subset)
                subset.pop()
        
        backtrack(0,0,[])
        return (ret)
        # return ret